### PR TITLE
Migrate sync demos from 01 to 03

### PR DESCRIPTION
**Pull request by peter-csala**
## Description:
- Renamed `Demo00_NoPolicy` to `Demo00_NoStrategy`
- Migrated form Demo01 to Demo03 under Sync folder to use `HttpClient` instead of `WebClient`
- Migrated policies to strategies

### CODE REVIEW

**Comment by peter-csala** [comment]
@martincostello What do you think would it make sense to mention the migration process on the readme.md?

**Comment by martincostello** [comment]
Yeah a temporary banner saying it's a work-in-progress is probably a good idea.

File: PollyDemos/Sync/Demo01_RetryNTimes.cs [hunk 1]
```diff
@@ -23,80 +22,87 @@ public class Demo01_RetryNTimes : SyncDemo
...
-                // Tell the user what they've won!
-                progress.Report(ProgressWithMessage("Policy logging: " + exception.Message, Color.Yellow));
-                retries++;
-            });
+                ShouldHandle = new PredicateBuilder().Handle<Exception>(),
+                MaxRetryAttempts = 3, // Retry up to 3 times
+                OnRetry = args =>
+                {
+                    // Due to fact how we have defined the ShouldHandle this delegate is called only if an exception occurred.
+                    // Please note the ! sign (null-forgiving operator) at the end of the command.
```
Comment (martincostello) [inline PollyDemos/Sync/Demo01_RetryNTimes.cs:50]:
```suggestion
                    // Due to how we have defined ShouldHandle, this delegate is called only if an exception occurred.
                    // Please note the ! sign (null-forgiving operator) at the end of the command.
```

File: PollyDemos/Sync/Demo01_RetryNTimes.cs [hunk 2]
```diff
@@ -23,80 +22,87 @@ public class Demo01_RetryNTimes : SyncDemo
...
-                var internalCancel = false;
-                totalRequests = 0;
-                // Do the following until a key is pressed
-                while (!internalCancel && !cancellationToken.IsCancellationRequested)
+                totalRequests++;
+
+                try
+                {
+                    // Retry the following call according to the strategy - 3 times.
+                    // the Execute() overload takes a CancellationToken, but the decorated code does not honour it.
```
Comment (martincostello) [inline PollyDemos/Sync/Demo01_RetryNTimes.cs:70]:
```suggestion
                    // The Execute() overload takes a CancellationToken, but the decorated code does not honour it.
```

File: PollyDemos/Sync/Demo01_RetryNTimes.cs [hunk 3]
```diff
@@ -23,80 +22,87 @@ public class Demo01_RetryNTimes : SyncDemo
...
+                    strategy.Execute(ct =>
+                        {
+                            // This code is executed within the strategy
+
+                            // Make a request and get a response
+                            var url = $"{Configuration.WEB_API_ROOT}/api/values/{totalRequests}";
+                            // Please note that the cancellation token is not used here.
```
Comment (martincostello) [inline PollyDemos/Sync/Demo01_RetryNTimes.cs:77]:
```suggestion
                            // Note that the cancellation token is not used here.
```

**Comment by peter-csala** [comment]
> Yeah a temporary banner saying it's a work-in-progress is probably a good idea.

Let me amend the current PR with that.

File: PollyDemos/Sync/Demo01_RetryNTimes.cs [hunk 4]
```diff
@@ -23,80 +22,87 @@ public class Demo01_RetryNTimes : SyncDemo
...
+                            var response = client.Send(new HttpRequestMessage(HttpMethod.Get, url));
+
+                            // Display the response message on the console
+                            // Please note that the cancellation token is not used here.
```
Comment (martincostello) [inline PollyDemos/Sync/Demo01_RetryNTimes.cs:81]:
```suggestion
                            // Display the response message on the console.
                            // Note that the cancellation token is not used here.
```

File: PollyDemos/Sync/Demo01_RetryNTimes.cs (hunk 3 above)
Comment (martincostello) [reply PollyDemos/Sync/Demo01_RetryNTimes.cs:77]: Also let's say _why_ it isn't being used.

File: PollyDemos/Sync/Demo02_WaitAndRetryNTimes.cs [hunk 5]
```diff
@@ -20,87 +19,92 @@ public class Demo02_WaitAndRetryNTimes : SyncDemo
...
-                // Do the following until a key is pressed
-                while (!internalCancel && !cancellationToken.IsCancellationRequested)
+                ShouldHandle = new PredicateBuilder().Handle<Exception>(),
+                MaxRetryAttempts = 3,
+                Delay = TimeSpan.FromMilliseconds(200),  // Wait 200ms between each try
+                OnRetry = args =>
                 {
-                    totalRequests++;
+                    // Due to fact how we have defined the ShouldHandle this delegate is called only if an exception occurred.
+                    // Please note the ! sign (null-forgiving operator) at the end of the command.
```
Comment (martincostello) [inline PollyDemos/Sync/Demo02_WaitAndRetryNTimes.cs:52]:
```suggestion
                    // Due to how we have defined ShouldHandle, this delegate is called only if an exception occurred.
                    // Note the ! sign (null-forgiving operator) at the end of the command.
```

File: PollyDemos/Sync/Demo03_WaitAndRetryNTimes_WithEnoughRetries.cs [hunk 6]
```diff
@@ -24,84 +23,90 @@ public class Demo03_WaitAndRetryNTimes_WithEnoughRetries : SyncDemo
...
-            // Define our policy:
-            var policy = Policy.Handle<Exception>().WaitAndRetry(
-                20, // Retry up to 20 times! - should be enough that we eventually succeed.
-                attempt => TimeSpan.FromMilliseconds(200), // Wait 200ms between each try.
-                (exception, calculatedWaitDuration) => // Capture some info for logging!
+            // Define our strategy:
+            var strategy = new ResiliencePipelineBuilder().AddRetry(new()
+            {
+                ShouldHandle = new PredicateBuilder().Handle<Exception>(),
+                MaxRetryAttempts = 20, // Retry up to 20 times! - should be enough that we eventually succeed.
```
Comment (martincostello) [inline PollyDemos/Sync/Demo03_WaitAndRetryNTimes_WithEnoughRetries.cs:47]:
```suggestion
                MaxRetryAttempts = 20, // Retry up to 20 times - this should be enough that we eventually succeed.
```

File: PollyDemos/Sync/Demo03_WaitAndRetryNTimes_WithEnoughRetries.cs [hunk 7]
```diff
@@ -24,84 +23,90 @@ public class Demo03_WaitAndRetryNTimes_WithEnoughRetries : SyncDemo
...
+                Delay = TimeSpan.FromMilliseconds(200),  // Wait 200ms between each try
+                OnRetry = args =>
                 {
-                    // This is your new exception handler! 
-                    // Tell the user what they've won!
-                    progress.Report(ProgressWithMessage("Policy logging: " + exception.Message, Color.Yellow));
+                    // Due to fact how we have defined the ShouldHandle this delegate is called only if an exception occurred.
+                    // Please note the ! sign (null-forgiving operator) at the end of the command.
```
Comment (martincostello) [inline PollyDemos/Sync/Demo03_WaitAndRetryNTimes_WithEnoughRetries.cs:52]:
```suggestion
                    // Due to how we have defined ShouldHandle, this delegate is called only if an exception occurred.
                    // Note the ! sign (null-forgiving operator) at the end of the command.
```

File: PollyDemos/Sync/Demo03_WaitAndRetryNTimes_WithEnoughRetries.cs [hunk 8]
```diff
@@ -24,84 +23,90 @@ public class Demo03_WaitAndRetryNTimes_WithEnoughRetries : SyncDemo
...
-                totalRequests = 0;
-                var internalCancel = false;
+                totalRequests++;
 
-                // Do the following until a key is pressed
-                while (!internalCancel && !cancellationToken.IsCancellationRequested)
+                try
+                {
+                    // Retry the following call according to the strategy - 20 times.
+                    // the Execute() overload takes a CancellationToken, but the decorated code does not honour it.
```
Comment (martincostello) [inline PollyDemos/Sync/Demo03_WaitAndRetryNTimes_WithEnoughRetries.cs:74]:
```suggestion
                    // The Execute() overload takes a CancellationToken, but the decorated code does not honour it.
```

File: PollyDemos/Sync/Demo03_WaitAndRetryNTimes_WithEnoughRetries.cs [hunk 9]
```diff
@@ -24,84 +23,90 @@ public class Demo03_WaitAndRetryNTimes_WithEnoughRetries : SyncDemo
...
+                        {
+                            // This code is executed within the strategy
+
+                            // Make a request and get a response
+                            var url = $"{Configuration.WEB_API_ROOT}/api/values/{totalRequests}";
+                            // Please note that the cancellation token is not used here.
+                            var response = client.Send(new HttpRequestMessage(HttpMethod.Get, url));
+
+                            // Display the response message on the console
+                            // Please note that the cancellation token is not used here.
```
Comment (martincostello) [inline PollyDemos/Sync/Demo03_WaitAndRetryNTimes_WithEnoughRetries.cs:86]:
```suggestion
                            // Nte that the cancellation token is not used here.
                            var response = client.Send(new HttpRequestMessage(HttpMethod.Get, url));

                            // Display the response message on the console.
                            // Note that the cancellation token is not used here.
```

File: README.md [hunk 10]
```diff
@@ -4,20 +4,26 @@
 
 Provides sample implementations using the [Polly library](https://www.github.com/App-vNext/Polly). The intent of this project is to help newcomers kick-start the use of [Polly](https://www.github.com/App-vNext/Polly) within their own projects.  The samples demonstrate the policies in action, against faulting endpoints
 
+> [!NOTE]
+> **Sample code migration to V8 is in progress!**
+>
+> As **Polly v8** has reached **feature-complete** state it's time to update the `Polly-Samples` as well.
+> The migration process happens gradually so, we are asking for your patience.
```
Comment (martincostello) [inline README.md:11]:
```suggestion
> The migration process happens gradually so, we are asking for your patience while we get everything updated.
```

File: PollyDemos/Sync/Demo01_RetryNTimes.cs (hunk 3 above)
Comment (peter-csala) [reply PollyDemos/Sync/Demo01_RetryNTimes.cs:77]: Let me double check what could be here the original intent.

I saw that the `CancellationToken` is honored only Demo 8 and 9

Comment (martincostello) [reply PollyDemos/Sync/Demo01_RetryNTimes.cs:77]: Yeah, I wasn't sure why either just from looking at it. I wondered if it was for pessimistic timeouts, but it's a retry sample.

Comment (peter-csala) [reply PollyDemos/Sync/Demo01_RetryNTimes.cs:77]: I did find any clue. Are you okay to change the samples to take advantage of the cancellation tokens?

Comment (martincostello) [reply PollyDemos/Sync/Demo01_RetryNTimes.cs:77]: That's OK with me.

**Review by martincostello — APPROVED** [review]

**Pull request merged**